src/runtime/opengl/opengl_module.cc

```diff
@@ -0,0 +1,225 @@
+  // resource handle
+  std::shared_ptr<ModuleNode> sptr_;
+  // The name of the function.
+  std::string func_name_;
+  // convert code for void argument
+  std::vector<size_t> arg_size_;
+  // thread axis config
+  ThreadAxisConfig thread_axis_cfg_;
+};
+
+OpenGLModuleNode::OpenGLModuleNode(
+    std::string data, std::string fmt,
+    std::unordered_map<std::string, FunctionInfo> fmap)
+    : workspace_(gl::OpenGLWorkspace::Global()), data_(std::move(data)),
+      fmt_(std::move(fmt)), fmap_(std::move(fmap)),
+      program_(workspace_->CreateProgram(data_.c_str())) {
+  LOG(INFO) << "OpenGLModuleNode(" << data << ", " << fmt << ", "
+                    << fmap.size() << ")";
+  CHECK(fmt_ == "gl") << "Unknown OpenGL format " << fmt_;
+}
+
+PackedFunc OpenGLModuleNode::GetFunction(
+    const std::string& name,
+    const std::shared_ptr<ModuleNode>& sptr_to_self) {
+  LOG(INFO) << "OpenGLModuleNode::GetFunction(" << name << ")";
+  CHECK_EQ(sptr_to_self.get(), this);
+  CHECK_NE(name, symbol::tvm_module_main) << "Device function do not have main";
+
+  auto it = fmap_.find(name);
+  if (it == fmap_.end()) return PackedFunc();
+  const FunctionInfo& info = it->second;
+
+  std::vector<size_t> arg_size(info.arg_types.size());
+  for (size_t i = 0; i < info.arg_types.size(); ++i) {
+    TVMType t = info.arg_types[i];
+    CHECK_EQ(t.lanes, 1U);
+    uint32_t bits = t.bits;
+    CHECK_EQ(bits % 8, 0U);
+    arg_size[i] = bits / 8;
+  }
+
+  // initialize the wrapped func.
+  OpenGLWrappedFunc f(this, sptr_to_self, name, arg_size,
+                      info.thread_axis_tags);
+  return PackFuncVoidAddr(f, info.arg_types);
+}
+
+std::string OpenGLModuleNode::GetSource(const std::string& format) {
+  if (format == fmt_) return data_;
+  if (fmt_ == "gl") {
+    return data_;
+  } else {
+    return "";
+  }
+}
+
+OpenGLWrappedFunc::OpenGLWrappedFunc(
+    OpenGLModuleNode* m,
+    std::shared_ptr<ModuleNode> sptr,
+    std::string func_name,
+    std::vector<size_t> arg_size,
+    const std::vector<std::string>& thread_axis_tags)
+    : m_(m), sptr_(std::move(sptr)), func_name_(std::move(func_name)),
+      arg_size_(std::move(arg_size)) {
+  LOG(INFO) << "OpenGLWrappedFunc(" << func_name_ << ", "
+            << "nargs = " << arg_size_.size() << ", "
+            << "nthread_axis_tags = " << thread_axis_tags.size() << ")";
+  for (auto& a : arg_size_) {
+    LOG(INFO) << a;
+  }
+  for (auto& t : thread_axis_tags) {
+    LOG(INFO) << t;
+  }
+
+  thread_axis_cfg_.Init(arg_size_.size(), thread_axis_tags);
+}
+
+void OpenGLWrappedFunc::operator()(TVMArgs args, TVMRetValue* rv,
+                                   void** void_args) const {
+  LOG(INFO) << "OpenGLWrappedFunc::operator()";
+
+  gl::Texture* arg0 = *static_cast<gl::Texture**>(void_args[0]);
+  std::vector<std::pair<std::string, gl::Texture*>> name_texture_pairs;
+  for (size_t i = 1; i != arg_size_.size(); ++i) {
+    name_texture_pairs.push_back(std::make_pair(
```

Comment:
NOTE, not necessary every argument is memory pointer. We could pass in other integer variables(integer types) as uniform

Reply:
Changed.